packages/mongodb-memory-server-core/src/util/MongoBinaryDownload.ts

```diff
@@ -353,126 +356,199 @@ export class MongoBinaryDownload {
   }
 
   /**
-   * Downlaod given httpOptions to tempDownloadLocation, then move it to downloadLocation
+   * Download given httpOptions to tempDownloadLocation, then move it to downloadLocation
+   * @param url The URL to download the file from
    * @param httpOptions The httpOptions directly passed to https.get
    * @param downloadLocation The location the File should be after the download
    * @param tempDownloadLocation The location the File should be while downloading
+   * @param maxRetries Maximum number of retries on download failure
+   * @param baseDelay Base delay in milliseconds for retrying the download
    */
   async httpDownload(
     url: URL,
     httpOptions: RequestOptions,
     downloadLocation: string,
-    tempDownloadLocation: string
+    tempDownloadLocation: string,
+    maxRetries?: number,
+    baseDelay: number = 1000
   ): Promise<string> {
     log('httpDownload');
     const downloadUrl = this.assignDownloadingURL(url);
 
-    const maxRedirects = parseInt(resolveConfig(ResolveConfigVariables.MAX_REDIRECTS) || '');
+    const maxRedirects = parseInt(resolveConfig(ResolveConfigVariables.MAX_REDIRECTS) ?? '');
     const useHttpsOptions: Parameters<typeof https.get>[1] = {
       maxRedirects: Number.isNaN(maxRedirects) ? 2 : maxRedirects,
       ...httpOptions,
     };
 
+    // Get maxRetries from config if not provided
+    const retriesFromConfig = parseInt(resolveConfig(ResolveConfigVariables.MAX_RETRIES) ?? '');
+    const retries =
+      typeof maxRetries === 'number'
+        ? maxRetries
+        : !Number.isNaN(retriesFromConfig)
+          ? retriesFromConfig
+          : 3;
+
+    for (let attempt = 0; attempt <= retries; attempt++) {
+      try {
+        return await this.attemptDownload(
+          url,
+          useHttpsOptions,
+          downloadLocation,
+          tempDownloadLocation,
+          downloadUrl,
+          httpOptions
+        );
+      } catch (error: any) {
+        const shouldRetry =
+          (error instanceof DownloadError &&
+            retryableStatusCodes.some((code) => error.message.includes(code.toString()))) ||
+          (error?.code && retryableErrorCodes.includes(error.code));
+
+        if (!shouldRetry || attempt === retries) {
+          throw error;
+        }
+
+        const base = baseDelay * Math.pow(2, attempt);
+        const jitter = Math.floor(Math.random() * 1000);
+        const delay = base + jitter;
+        log(
+          `httpDownload: attempt ${attempt + 1} failed with ${error.message}, retrying in ${delay}ms...`
+        );
+        await new Promise((resolve) => setTimeout(resolve, delay));
+      }
+    }
+
+    throw new Error('Max retries exceeded');
```

Comment:
Shouldnt that also be a `DownloadError`?

Reply:
Only retryable errors are DownloadError

Reply:
No, a `DownloadError` is when downloading fails, which it does if max retries are exceeded.
Also see:
https://github.com/typegoose/mongodb-memory-server/blob/e4992fd4f89d5025d315d05d7ca6d43358cf62be/packages/mongodb-memory-server-core/src/util/errors.ts#L201-L204